Add Spi support

folkertdev opened the pull request:
This PR adds SPI communication support.

This work was done at @tweedegolf while working on a commercial project, and we'd now like open source it. It currently runs in production.

Clearly, adding a whole new communication protocol is a pretty invasive change. Here I'm showing the version that we ended up using in our projects, but that does not mean it's the best approach for `lis3dh-rs`  in general. We're open to modifying this code to make it fit.

At a high level

- common functionality is put in the `Lis3dhImpl` trait. For all but 3 functions there are default implementations. This allows us to share a lot of code between the I2C and SPI implementations
- i2c and spi have their own modules, with their own `Lis3dh` struct and implementations of the `Lis3dhImpl` trait and accelerometer traits. 
- The `Error` type needs 2 type paramters now, one for bus and one for pin errors.

Some open questions

- the `Lis3dhImpl` trait leaks more implementation details than the current version. Maybe that's fine, maybe we should not expose that trait but make a supertrait that only exposes the desired methods.
- testing with `embedded_hal_mock`
- there may be some tricks that can give us just one `Lis3dh` struct with `new_i2c()` and `new_spi` functions. Not sure if that's actually nicer in practice though

folkertdev committed:
clippy

folkertdev committed:
move implementation into trait

folkertdev committed:
move i2c implementation to its own file

folkertdev committed:
add with_config method

folkertdev committed:
add spi code

folkertdev committed:
centralize accelerometer impl

folkertdev committed:
re-export the slave address from i2c module

folkertdev committed:
remove delays

folkertdev committed:
remove delay type var

folkertdev committed:
refactor into core trait and structs

sajattack commented on src/lib.rs:
```diff
@@ -372,37 +383,37 @@ where
+    pub enable_temperature: bool,
+    /// Whether data should only be updated after reading
+    pub block_data_update: bool,
+    pub mode: Mode,
```
Please add doc comments for the rest of these.

sajattack commented:
Looks great, thanks for contributing back :smile: Just needs a bit more documentation I think.

sajattack commented:
@jacobrosenthal @BenBergman would either of you like to review this as well?

sajattack commented on src/i2c.rs:
```diff
@@ -0,0 +1,130 @@
+
+    /// Read from the registers for each of the 3 axes.
+    fn read_accel_bytes(&mut self) -> Result<[u8; 6], Error<Self::BusError, Self::PinError>> {
+        let mut data = [0u8; 6];
+
+        self.i2c
+            .write_read(self.address, &[Register::OUT_X_L.addr() | 0x80], &mut data)
+            .map_err(Error::Bus)
+            .and(Ok(data))
+    }
+
+    /// Write a byte to the given register.
+    fn write_register(
+        &mut self,
+        register: Register,
+        value: u8,
+    ) -> Result<(), Error<Self::BusError, Self::PinError>> {
+        if register.read_only() {
+            return Err(Error::WriteToReadOnly);
+        }
+
+        self.i2c
+            .write(self.address, &[register.addr(), value])
+            .map_err(Error::Bus)
+    }
+
+    /// Read a byte from the given register.
+    fn read_register(
+        &mut self,
+        register: Register,
+    ) -> Result<u8, Error<Self::BusError, Self::PinError>> {
+        let mut data = [0];
+
+        self.i2c
+            .write_read(self.address, &[register.addr()], &mut data)
+            .map_err(Error::Bus)
+            .and(Ok(data[0]))
+    }
+}
+
```
Are these exposed now with your changes?

sajattack replied:
Maybe make a trait called Lis3dhRaw or somesuch to hide these.

folkertdev committed:
add spi

folkertdev commented:
yes those are now exposed. That's why I kept playing around a bit, and I found this alternative solution that uses just one `Lis3dh` struct, and you can pick either spi or i2c based on using `new_i2c` or `new_spi`. I personally like this design better exactly because it gives better control about what is exposed, and because everything can be nicely exposed from just one module. 

[the alternative implementation](https://github.com/BenBergman/lis3dh-rs/compare/master...folkertdev:one-struct-spi)

In this case, there is the `Lis3dhCore` trait (though `Raw` might be a better name) that gets implemented by a struct wrapping an I2C device, and by a struct wrapping an SPI device. Then the main `Lis3dh` struct demands an argument that implements this `Core` trait. in its constructors. 

Downside is that the error type is now always an associated type, so it's harder to figure out what exactly is in an error.

What are your thoughts on the tradeoffs here?

sajattack commented:
Sorry, I don't understand what you mean by this
> Downside is that the error type is now always an associated type, so it's harder to figure out what exactly is in an error.

folkertdev commented:
actually, nevermind. What I thought was a problem is really just more of what this library already has. It's generic over the error type that the I2C device may have. My worry was that it would be hard to figure out what type the error has (and usually, at least on the devices that I work on, a `Debug` implementation will not save you. I generally need `defmt::Format`). 

In this PR, instead we have errors of the form `Error<Self::BusError, Self::PinError>`. What type `BusError` has is not immediately clear, especially when the `Lis3dhCore` trait is not publicly exposed. But I guess this is not too different from not really knowing the concrete I2C error type.

folkertdev committed:
add new_spi function

folkertdev committed:
clippy

folkertdev committed:
add examples

folkertdev committed:
add Config type

folkertdev committed:
Revert "remove delay type var"

This reverts commit d03cc94b0df36d26febdaca07e667edfff58c044.

folkertdev committed:
Revert "remove delays"

This reverts commit b724c5aaa1a220c6b97872ab1302a02454daa802.

folkertdev committed:
Revert "re-export the slave address from i2c module"

This reverts commit c3b8fb687878ffe27a4f0005c13b8ce603e898a9.

folkertdev committed:
Revert "centralize accelerometer impl"

This reverts commit 8a8c508072635dc1de362207e651ca8fce31978d.

folkertdev committed:
Revert "add spi code"

This reverts commit 62f627a32ea1e2a96b32f8b8aa922133855ec086.

folkertdev committed:
Revert "add with_config method"

This reverts commit 9626ff7fc131aebddd1e130045abc87e5b209500.

folkertdev committed:
Revert "move i2c implementation to its own file"

This reverts commit 078c5ae6a98a5d3c18c6dfef554006fa2b18fbfb.

folkertdev committed:
Revert "move implementation into trait"

This reverts commit 10131035c470ff06b0a3569e4b01cc35e076bd6e.

folkertdev committed:
Revert "clippy"

This reverts commit 222ee48c4cc7ec52fc4244504b1db52263aca510.

folkertdev commented:
OK, I moved this over to the new, better, one-file approach, added more doc comments, and even some full configuration examples. Let me know what you think.

We also have interrupt configuration working, which I can add as a separate PR.

sajattack commented:
Remind me if I haven't touched this by Monday, I want to take some time to try it out.

sajattack commented:
Seems like i2c still works anyway, don't have spi to test with :+1:

sajattack commented:
Published as crate version 0.4.1. I accidentally published 0.4.0 with no changes since 0.3.0 and I don't think I have a way to fix that. I pulled the wrong git remote.
